staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -601,6 +672,46 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
 	return info.storages[info.storageVersion].CustomResource, nil
 }
 
+// getOrCreateServingInfoForWrite is getOrCreateServingInfoFor, but blocks CR
+// write requests until the storage version update is processed, aborted, or
+// timed-out.
+// If the storage version update is processed, unblock and allow the write request.
+// If the storage version update is aborted, unblock and reject the write request.
+// If the storage version update timed-out, unblock and allow the write request.
+func (r *crdHandler) getOrCreateServingInfoForWrite(uid types.UID, name string) (*crdInfo, error) {
+	ret, err := r.getOrCreateServingInfoFor(uid, name)
+
+	// Double check the feature gate is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) &&
+		ret.storageVersionUpdateProcessedCh != nil {
+		// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+		// to register themselves to the wait group. Ideally the storage version update should
+		// not cause the requests to miss the 1s window; otherwise the requests may
+		// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+		// first CR request establishes the underlying storage).
+		select {
+		case <-ret.storageVersionUpdateProcessedCh:
+		case <-ret.storageVersionUpdateAbortedCh:
+			// Storage version update failed because the CRD was deleted. Here
+			// we return a 404 to fail this in-flight request.
```

Comment:
I believe this must be a 403, not a 404. Please double check. Afaik for the garbage collector it makes a difference: 404 = gone. 403 = unknown owner reference. @deads2k can comment.

Reply:
> 403 StatusForbidden
Indicates that the server can be reached and understood the request, but refuses to take any further action, because it is configured to deny access for some reason to the requested resource by the client.
Suggested client recovery behavior: Do not retry. Fix the request.

I agree 404 is better than 403. 404 allows us to provide more detail in the error message about why the request got rejected. Plus the server truly did understand the request. I changed to 403

I'm not so sure about the recovery behavior though. There isn't much the CR client can do to fix the request (assuming a different user deleted the CRD). A 503 can let the CR client to simply retry, and it will get either a 404 (CRD teardown finished) or a success (CRD re-created) later.